File: src/WirePact.PoC.Translator/Services/Translator.cs
using System.Threading.Tasks;
using Envoy.Config.Core.V3;
using Envoy.Service.Auth.V3;
using Envoy.Type.V3;
using Google.Rpc;
using Grpc.Core;
using Microsoft.Extensions.Logging;
using WirePact.PoC.Translator.Kubernetes;
using StatusCode = Envoy.Type.V3.StatusCode;

namespace WirePact.PoC.Translator.Services
{
    public class Translator : Authorization.AuthorizationBase
    {
        private readonly ILogger<Translator> _logger;
        private readonly ZitadelValidator _validator;
        private readonly CredentialsReader _reader;

        public Translator(ILogger<Translator> logger, ZitadelValidator validator, CredentialsReader reader)
        {
            _logger = logger;
            _validator = validator;
            _reader = reader;
        }

        public override async Task<CheckResponse> Check(CheckRequest request, ServerCallContext context)
        {
            _logger.LogDebug(
                $"Check Request To: {request.Attributes.Request.Http.Host} {request.Attributes.Request.Http.Path}");

            if (!request.Attributes.Request.Http.Headers.TryGetValue("authorization", out var authHeader))
            {
                _logger.LogWarning(
                    "No Authorization Header provided for request to: {host}{path}",
                    request.Attributes.Request.Http.Host,
                    request.Attributes.Request.Http.Path);
                return Fail(Code.FailedPrecondition, StatusCode.Unauthorized, "No Authorization Header provided.");
            }

            if (!await _validator.IsValidToken(authHeader.Replace("Bearer ", string.Empty)))
            {
                _logger.LogWarning(
                    "Zitadel rejected the token for request to: {host}{path}",
                    request.Attributes.Request.Http.Host,
                    request.Attributes.Request.Http.Path);
                return Fail(Code.Unauthenticated, StatusCode.Forbidden, "No valid Authorization provided");
            }

            return Success(_reader.BasicAuthCredentials);
        }

        private static CheckResponse Fail(Code code, StatusCode status, string body) => new()
        {
            Status = new() { Code = (int)code },
            DeniedResponse = new()
            {
                Status = new HttpStatus { Code = status },
                Body = body,
            },
        };

        private static CheckResponse Success(string basicAuth) => new()
        {
            Status = new() { Code = (int)Code.Ok },
            OkResponse = new()
            {
                Headers =
                {
                    new HeaderValueOption
                    {
                        Header = new()
                        {
                            Key = "Authorization",
                            Value = $"Basic {basicAuth}",
                        },
                    },
                },
            },
        };
    }
}

File: src/WirePact.PoC.Translator/Translator.fs
namespace WirePact.PoC.Translator

open Envoy.Config.Core.V3
open Envoy.Service.Auth.V3
open Envoy.Type.V3
open FSharp.Control.Tasks
open Google.Rpc
open Microsoft.Extensions.Logging

type Translator(logger: ILogger<Translator>, reader: CredentialReader, validator: ZitadelValidator) =
    inherit Authorization.AuthorizationBase()

    let fail (code: Code) status message =
        CheckResponse
            (Status = Status(Code = int code),
             DeniedResponse = DeniedHttpResponse(Status = HttpStatus(Code = status), Body = message))

    let success basicCredentials =
        let okResponse = OkHttpResponse()

        okResponse.Headers.Add
            (HeaderValueOption(Header = HeaderValue(Key = "Authorization", Value = $"Basic {basicCredentials}")))

        CheckResponse(Status = Status(Code = int Code.Ok), OkResponse = okResponse)

    override _.Check(request, _) =
        task {
            logger.LogDebug
                $"Check Request To: {request.Attributes.Request.Http.Host} {request.Attributes.Request.Http.Path}"

            let (hasHeader, authHeader) =
                request.Attributes.Request.Http.Headers.TryGetValue "authorization"

            if not hasHeader then
                logger.LogWarning
                    ("No Authorization Header provided for request to: {host}{path}",
                     request.Attributes.Request.Http.Host,
                     request.Attributes.Request.Http.Path)

                return fail Code.FailedPrecondition StatusCode.Unauthorized "No Authorization Header provided."
            else
                let! validToken = validator.IsValidToken(authHeader.Replace("Bearer ", ""))

                if validToken then
                    return success reader.BasicAuthCredentials
                else
                    logger.LogWarning
                        ("Zitadel rejected the token for request to: {host}{path}",
                         request.Attributes.Request.Http.Host,
                         request.Attributes.Request.Http.Path)

                    return fail Code.Unauthenticated StatusCode.Forbidden "No valid Authorization provided"
        }
